posts/2023-05-17-advanced-kotlin.adoc

```diff
@@ -0,0 +1,326 @@
+
+However, Kotlin imposes stricter rules on variance to ensure type safety. In Kotlin, a type parameter marked as _out_ is only available in output positions (e.g., as the return type of a function) but not in input positions (e.g., as function parameters).
+Similarly, an _in_ parameter is only allowed in input positions. This safety measure helps prevent potential runtime _ClassCastException_ errors.
+
+Here's an example of using out (covariant) keyword in Kotlin:
+
+[source,kotlin]
+----
+abstract class MyProducer<out T> {
+    abstract fun produce(): T
+}
+
+fun demo(str: MyProducer<String>) {
+    val objects: MyProducer<Any> = str // This is OK, since T is an out-parameter
+    // ...
+}
+----
+
+In this example, T is declared as an out parameter, so MyProducer<String> can be treated as MyProducer<Any>, respecting the subtyping relationship between String and Any.
+
+And an example for in (contravariant):
+
+[source,kotlin]
+----
+abstract class MyConsumer<in T> {
+    abstract fun consume(item: T)
+}
+
+fun demo(ints: MyConsumer<Int>) {
+    val anyConsumer: MyConsumer<Any> = ints // Error: Type mismatch.
+    val numberConsumer: MyConsumer<Number> = ints // This is OK, since T is an in-parameter
+    // ...
+}
+----
+Here, T is declared as an in parameter, meaning MyConsumer<Int> can be treated as MyConsumer<Number>, again preserving the subtyping relationship.
+
+==== Type Projections
+In Kotlin, generic classes have type parameters which may come with variance annotations (like _in_ or _out_).
+However, not all classes can be (or should be) strictly defined as covariant or contravariant.
+There are cases where a class works with its type parameter in both _in_ and _out_ positions.
+
+This is where type projections come in. They let us define the variance for type parameters "on the fly", for specific instances of the class.
+This allows us to work with such classes in a more flexible and safer way.
+For example:
+
+[source,kotlin]
+----
+class MyMutableList<T> {
+    fun add(item: T) { /*...*/ }
+    fun get(index: Int): T { /*...*/ }
+}
+
+fun printFirst(myList: MyMutableList<out Any>) {
+    println(myList.get(0))
+}
+
+fun addToEnd(myList: MyMutableList<in String>, item: String) {
+    myList.add(item)
+}
+----
+In this example, MyMutableList works with T in both _in_ and _out_ positions, so we can't mark T as _in_ or _out_ in its declaration.
+Instead, we can use type projections in the functions _printFirst_ and _addToEnd_.
+
+Type projections in Kotlin serve a similar purpose as use-site variance in Java but are absent in Scala.
+
+==== Comparing with Scala's Type System
+While both Kotlin and Scala have powerful type systems, they each have unique features and handle some concepts differently.
+Scala's type system is more complex and powerful, with features such as higher-kinded types and implicit parameters, Scala can express complex abstractions that are not possible in Kotlin.
```

Comment:
I would suggest a period after parameters and start the sentence anew with Scala